rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/SpatialMatcher.scala

```diff
@@ -450,43 +446,33 @@ trait SpatialMatcherInstances {
 
   implicit def bundleSpatialMatcherInstance[F[_]: Splittable: Alternative: Monad: _error]
       : SpatialMatcher[F, Bundle, Bundle] =
-    (target, pattern) => {
-      if (pattern == target)
-        ().pure[F]
-      else {
-        MonoidK[F].empty[Unit]
-      }
-    }
+    (target, pattern) => Alternative_[F].guard(pattern == target)
 
   implicit def sendSpatialMatcherInstance[F[_]: Splittable: Alternative: Monad: _error: _freeMap: _short]
       : SpatialMatcher[F, Send, Send] =
     (target, pattern) =>
-      if (target.persistent != pattern.persistent)
-        MonoidK[F].empty[Unit]
-      else
-        for {
-          _ <- spatialMatch(target.chan, pattern.chan)
-          _ <- foldMatch(target.data, pattern.data)
-        } yield ()
+      for {
+        _ <- Alternative_[F].guard(target.persistent == pattern.persistent)
+        _ <- spatialMatch(target.chan, pattern.chan)
+        _ <- foldMatch(target.data, pattern.data)
+      } yield ()
 
   implicit def receiveSpatialMatcherInstance[F[_]: Splittable: Alternative: Monad: _error: _freeMap: _short]
       : SpatialMatcher[F, Receive, Receive] =
     (target, pattern) =>
-      if (target.persistent != pattern.persistent)
-        MonoidK[F].empty[Unit]
-      else
-        for {
-          _ <- listMatchSingle(target.binds, pattern.binds)
-          _ <- spatialMatch(target.body, pattern.body)
-        } yield ()
+      for {
+        _ <- Alternative_[F].guard(target.persistent == pattern.persistent)
+        _ <- listMatchSingle(target.binds, pattern.binds)
+        _ <- spatialMatch(target.body, pattern.body)
+      } yield ()
```

Comment:
So `listMatchSingle` doesn't get executed if `target.persistent != pattern.persistent`?

Reply:
Yes. Alternative.guard(false) == MonoidK.empty, and that short-circuits a monadic bind (flatMap). (I think the laws-based explanation would be sth. like "whatever ap empty == empty due to Alternative laws, and because Monad must be consistent with Applicative (which Alternative is), empty flatMap whatever == empty" - but can't tell for sure)

Notice Altierntive_ is just a wrapper over cats.Alternative and the `guard` method comes from there. (In general whenever we have Foo_, this is just a wrapper over cats.Foo, just to avoid 'ambiguous implicit' errors. Other examples: ApplicativeError_, MonadError_).